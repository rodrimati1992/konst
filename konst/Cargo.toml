[package]
name = "konst"
version = "0.3.5"
authors = ["rodrimati1992 <rodrimatt1985@gmail.com>"]
rust-version = "1.65.0"
edition = "2021"
license = "Zlib"
description = "Const equivalents of std functions, compile-time comparison, and parsing"
documentation = "https://docs.rs/konst/"
readme="../README.md"
keywords = ["no-std", "const_fn", "parsing"]
categories = ["no-std", "parsing"]
repository = "https://github.com/rodrimati1992/konst/"
include = [
    "Cargo.toml", 
    "src/**/*.rs", 
    "../README.md",
    "LICENSE-ZLIB.md", 
]


[dependencies.konst_proc_macros]
version = "=0.3.0"
path = "../konst_proc_macros"
optional = true

[dependencies.konst_kernel]
version = "=0.3.5"
path = "../konst_kernel"
features = ["rust_1_64", "__for_konst"]

[dev-dependencies.rand]
version = "0.8.4"
default-features = false
features = ["small_rng"]

# dev-dependencies can't be optional
[dependencies.trybuild]
version = "1.0"
optional = true

[dependencies.typewit]
version = "1.1"
features = ["rust_1_61"]

[dependencies.const_panic]
version = "0.2.6"
features = ["rust_1_64"]
default-features = false

[features]
default = ["cmp", "iter", "parsing_proc"]


# Enables extra checks for debug
debug = ["konst_kernel/debug"]
cmp = []
iter = ["konst_kernel/iter"]
parsing_proc = ["parsing", "konst_proc_macros"]
parsing = []
rust_latest_stable = []
mut_refs = ["rust_latest_stable", "konst_kernel/mut_refs"]
nightly_mut_refs = ["mut_refs", "konst_kernel/nightly_mut_refs"]

alloc = ["konst_kernel/alloc"]

# Enables ui tests, which are intended for the latest stable release.
# This is not enabled in CI, because it's way too sensitive to 
# changes in how rustc formats errors
__ui = ["trybuild", "rust_latest_stable"]
__cp_derive = ["const_panic/derive", "const_panic/non_basic"]

docsrs = ["konst_kernel/docsrs"]



[package.metadata.docs.rs]
features = ["docsrs", "alloc", "cmp", "parsing_proc", "rust_latest_stable", "nightly_mut_refs"]

