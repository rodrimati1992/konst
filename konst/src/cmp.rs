//! Comparisong-related items.

mod cmp_wrapper;

mod const_cmp;

pub use self::{
    cmp_wrapper::CmpWrapper,
    const_cmp::{ConstCmp, ConstCmpUnref, IsAConstCmp},
};

#[doc(no_inline)]
pub use crate::polymorphism::kinds::{IsNotStdKind, IsRefKind, IsStdKind};
