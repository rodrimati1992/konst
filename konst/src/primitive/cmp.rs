use core::cmp::Ordering;

macro_rules! declare_int_cmp_fn {
    (
        ($type:ty, ($cmp_fn_name:ident))
        docs( $docs_eq:expr, $docs_cmp:expr, )
    ) => {
        crate::__delegate_const_ord! {
            #[doc = $docs_cmp]
            pub const fn $cmp_fn_name(copy left: $type, right: $type) -> Ordering {
                cmp_int!(left, right)
            }
        }
    };
}

__declare_fns_with_docs! {
    (u8, (cmp_u8))
    (u16, (cmp_u16))
    (u32, (cmp_u32))
    (u64, (cmp_u64))
    (u128, (cmp_u128))
    (usize, (cmp_usize))

    (i8, (cmp_i8))
    (i16, (cmp_i16))
    (i32, (cmp_i32))
    (i64, (cmp_i64))
    (i128, (cmp_i128))
    (isize, (cmp_isize))

    (bool, (cmp_bool))
    (char, (cmp_char))

    docs(default)

    macro = declare_int_cmp_fn!(),
}

__declare_fns_with_docs! {
    (Option<u8>, (eq_option_u8, cmp_option_u8))
    (Option<i8>, (eq_option_i8, cmp_option_i8))
    (Option<u16>, (eq_option_u16, cmp_option_u16))
    (Option<i16>, (eq_option_i16, cmp_option_i16))
    (Option<u32>, (eq_option_u32, cmp_option_u32))
    (Option<i32>, (eq_option_i32, cmp_option_i32))
    (Option<u64>, (eq_option_u64, cmp_option_u64))
    (Option<i64>, (eq_option_i64, cmp_option_i64))
    (Option<u128>, (eq_option_u128, cmp_option_u128))
    (Option<i128>, (eq_option_i128, cmp_option_i128))
    (Option<usize>, (eq_option_usize, cmp_option_usize))
    (Option<isize>, (eq_option_isize, cmp_option_isize))
    (Option<bool>, (eq_option_bool, cmp_option_bool))
    (Option<char>, (eq_option_char, cmp_option_char))

    docs(default)

    macro = __impl_option_cmp_fns!(
        params(l, r)
        eq_comparison = l == r,
        cmp_comparison = cmp_int!(l, r),
        parameter_copyability = copy,
    ),
}
