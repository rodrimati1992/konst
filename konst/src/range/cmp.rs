use core::ops::{Range, RangeInclusive};

use crate::cmp::{CmpWrapper, ConstCmp, IsAConstCmp, IsStdKind};

macro_rules! shared_range_impls {
    ($type:ty, $eq_fn_name:ident) => {
        impl ConstCmp for $type {
            type Kind = IsStdKind;
        }
        impl CmpWrapper<$type> {
            /// Compares `self` and `other` for equality.
            #[inline(always)]
            pub const fn const_eq(&self, other: &$type) -> bool {
                $eq_fn_name(&self.0, other)
            }
        }
    };
}

macro_rules! declare_range_cmp_fns {
    (
        ($type:ty, ($eq_fn_name:ident))

        docs( $docs_eq:expr, $docs_cmp:expr,)
    ) => {
        shared_range_impls! {$type, $eq_fn_name}

        impl<T> IsAConstCmp<IsStdKind, $type, T> {
            ///
            #[inline(always)]
            pub const fn coerce(self, range: &$type) -> CmpWrapper<$type> {
                CmpWrapper(Range {
                    start: range.start,
                    end: range.end,
                })
            }
        }

        __delegate_const_eq! {
            skip_coerce;
            for['a,]

            #[doc = $docs_eq]
            pub const fn $eq_fn_name(ref left: &'a $type, right: &'a $type) -> bool {
                left.start == right.start && left.end == right.end
            }
        }
    };
}

__declare_fns_with_docs! {
    (Range<u8>,    (eq_range_u8))
    (Range<u16>,   (eq_range_u16))
    (Range<u32>,   (eq_range_u32))
    (Range<u64>,   (eq_range_u64))
    (Range<u128>,  (eq_range_u128))
    (Range<usize>, (eq_range_usize))

    (Range<char>, (eq_range_char))

    docs(default)

    macro = declare_range_cmp_fns!(),
}

macro_rules! declare_rangeinclusive_cmp_fns {
    (
        ($type:ty, ($eq_fn_name:ident))

        docs( $docs_eq:expr, $docs_cmp:expr,)
    ) => {
        shared_range_impls! {$type, $eq_fn_name}

        impl<T> IsAConstCmp<IsStdKind, $type, T> {
            ///
            #[inline(always)]
            pub const fn coerce(self, range: &$type) -> CmpWrapper<$type> {
                CmpWrapper(RangeInclusive::new(*range.start(), *range.end()))
            }
        }

        #[doc = $docs_eq]
        pub const fn $eq_fn_name(left: &$type, right: &$type) -> bool {
            *left.start() == *right.start() && *left.end() == *right.end()
        }
    };
}

__declare_fns_with_docs! {
    (RangeInclusive<u8>,    (eq_rangeinc_u8,))
    (RangeInclusive<u16>,   (eq_rangeinc_u16))
    (RangeInclusive<u32>,   (eq_rangeinc_u32))
    (RangeInclusive<u64>,   (eq_rangeinc_u64))
    (RangeInclusive<u128>,  (eq_rangeinc_u128))
    (RangeInclusive<usize>, (eq_rangeinc_usize))

    (RangeInclusive<char>, (eq_rangeinc_char))

    docs(default)

    macro = declare_rangeinclusive_cmp_fns!(),
}
