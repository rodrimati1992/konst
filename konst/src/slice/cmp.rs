#[doc(inline)]
pub use crate::slice::{
    cmp_bytes as cmp_slice_u8, cmp_option_bytes as cmp_option_slice_u8, eq_bytes as eq_slice_u8,
    eq_option_bytes as eq_option_slice_u8,
};

use core::cmp::Ordering;

__declare_slice_cmp_fns! {
    import_path = "konst",

    (,,, u16, eq_slice_u16, cmp_slice_u16,)
    (,,, u32, eq_slice_u32, cmp_slice_u32,)
    (,,, u64, eq_slice_u64, cmp_slice_u64,)
    (,,, u128, eq_slice_u128, cmp_slice_u128,)
    (,,, usize, eq_slice_usize, cmp_slice_usize,)

    (,,, i8, eq_slice_i8, cmp_slice_i8,)
    (,,, i16, eq_slice_i16, cmp_slice_i16,)
    (,,, i32, eq_slice_i32, cmp_slice_i32,)
    (,,, i64, eq_slice_i64, cmp_slice_i64,)
    (,,, i128, eq_slice_i128, cmp_slice_i128,)
    (,,, isize, eq_slice_isize, cmp_slice_isize,)

    (,,, bool, eq_slice_bool, cmp_slice_bool,)
    (,,, char, eq_slice_char, cmp_slice_char,)
}

__delegate_const_eq! {
    skip_coerce;

    /// Compares two `&[&str]` for equality.
    pub const fn eq_slice_str(ref l: &[&str], r: &[&str]) -> bool {
        crate::const_eq_for!(slice; l, r, crate::eq_str)
    }
}

__delegate_const_ord! {
    skip_coerce;

    /// Compares two `&[&str]`, returning the order of `left` relative to `right`.
    pub const fn cmp_slice_str(ref left: &[&str], right: &[&str]) -> Ordering {
        crate::const_cmp_for!(slice; left, right, crate::cmp_str)
    }
}

__delegate_const_eq! {
    skip_coerce;

    /// Compares two `&[&[u8]]` for equality.
    pub const fn eq_slice_bytes(ref l: &[&[u8]], r: &[&[u8]]) -> bool {
        crate::const_eq_for!(slice; l, r, eq_slice_u8)
    }
}

__delegate_const_ord! {
    skip_coerce;

    /// Compares two `&[&[u8]]`, returning the order of `left` relative to `right`.
    pub const fn cmp_slice_bytes(ref left: &[&[u8]], right: &[&[u8]]) -> Ordering {
        crate::const_cmp_for!(slice; left, right, cmp_slice_u8)
    }
}

__declare_fns_with_docs! {
    (Option<&'a [u16]>, (eq_option_slice_u16, cmp_option_slice_u16))
    (Option<&'a [u32]>, (eq_option_slice_u32, cmp_option_slice_u32))
    (Option<&'a [u64]>, (eq_option_slice_u64, cmp_option_slice_u64))
    (Option<&'a [u128]>, (eq_option_slice_u128, cmp_option_slice_u128))
    (Option<&'a [usize]>, (eq_option_slice_usize, cmp_option_slice_usize))
    (Option<&'a [i8]>, (eq_option_slice_i8, cmp_option_slice_i8))
    (Option<&'a [i16]>, (eq_option_slice_i16, cmp_option_slice_i16))
    (Option<&'a [i32]>, (eq_option_slice_i32, cmp_option_slice_i32))
    (Option<&'a [i64]>, (eq_option_slice_i64, cmp_option_slice_i64))
    (Option<&'a [i128]>, (eq_option_slice_i128, cmp_option_slice_i128))
    (Option<&'a [isize]>, (eq_option_slice_isize, cmp_option_slice_isize))
    (Option<&'a [bool]>, (eq_option_slice_bool, cmp_option_slice_bool))
    (Option<&'a [char]>, (eq_option_slice_char, cmp_option_slice_char))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a,]
        params(l, r)
        eq_comparison = crate::cmp::CmpWrapper(l).const_eq(r),
        cmp_comparison = crate::cmp::CmpWrapper(l).const_cmp(r),
        parameter_copyability = copy,
    ),
}

__declare_fns_with_docs! {
    (Option<&'a [&'b str]>, (eq_option_slice_str, cmp_option_slice_str))
    (Option<&'a [&'b [u8]]>, (eq_option_slice_bytes, cmp_option_slice_bytes))

    docs(default)

    macro = __impl_option_cmp_fns!(
        for['a, 'b,]
        params(l, r)
        eq_comparison = crate::cmp::CmpWrapper(l).const_eq(r),
        cmp_comparison = crate::cmp::CmpWrapper(l).const_cmp(r),
        parameter_copyability = copy,
    ),
}
