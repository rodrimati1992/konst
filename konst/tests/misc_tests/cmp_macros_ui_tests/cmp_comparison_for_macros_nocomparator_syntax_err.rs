use konst::cmp;

use std::cmp::Ordering as Orde;


const _: bool = cmp::const_eq_for!();
const _: bool = cmp::const_ne_for!();
const _: Orde = cmp::const_cmp_for!();
const _: bool = cmp::const_lt_for!();
const _: bool = cmp::const_le_for!();
const _: bool = cmp::const_gt_for!();
const _: bool = cmp::const_ge_for!();

const _: bool = cmp::const_eq_for!(;);
const _: bool = cmp::const_ne_for!(;);
const _: Orde = cmp::const_cmp_for!(;);
const _: bool = cmp::const_lt_for!(;);
const _: bool = cmp::const_le_for!(;);
const _: bool = cmp::const_gt_for!(;);
const _: bool = cmp::const_ge_for!(;);

const _: bool = cmp::const_eq_for!(abcd;);
const _: bool = cmp::const_ne_for!(abcd;);
const _: Orde = cmp::const_cmp_for!(abcd;);
const _: bool = cmp::const_lt_for!(abcd;);
const _: bool = cmp::const_le_for!(abcd;);
const _: bool = cmp::const_gt_for!(abcd;);
const _: bool = cmp::const_ge_for!(abcd;);

const _: bool = cmp::const_eq_for!(slice; [0u8]);
const _: bool = cmp::const_ne_for!(slice; [0u8]);
const _: Orde = cmp::const_cmp_for!(slice; [0u8]);
const _: bool = cmp::const_lt_for!(slice; [0u8]);
const _: bool = cmp::const_le_for!(slice; [0u8]);
const _: bool = cmp::const_gt_for!(slice; [0u8]);
const _: bool = cmp::const_ge_for!(slice; [0u8]);

const _: bool = cmp::const_eq_for!(slice; [0u8],);
const _: bool = cmp::const_ne_for!(slice; [0u8],);
const _: Orde = cmp::const_cmp_for!(slice; [0u8],);
const _: bool = cmp::const_lt_for!(slice; [0u8],);
const _: bool = cmp::const_le_for!(slice; [0u8],);
const _: bool = cmp::const_gt_for!(slice; [0u8],);
const _: bool = cmp::const_ge_for!(slice; [0u8],);

