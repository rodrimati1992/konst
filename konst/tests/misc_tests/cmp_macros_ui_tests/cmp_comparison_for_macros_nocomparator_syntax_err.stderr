error: expected type argument, passed ``; valid type arguments:
       - slice
       - option
       - range
       - range_inclusive
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:6:17
  |
6 | const _: bool = cmp::const_eq_for!();
  |                 ^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `cmp::const_eq_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed ``; valid type arguments:
       - slice
       - option
       - range
       - range_inclusive
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:7:17
  |
7 | const _: bool = cmp::const_ne_for!();
  |                 ^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `$crate::cmp::const_eq_for` which comes from the expansion of the macro `cmp::const_ne_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed ``; valid type arguments:
       - slice
       - option
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:8:17
  |
8 | const _: Orde = cmp::const_cmp_for!();
  |                 ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `cmp::const_cmp_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed ``; valid type arguments:
       - slice
       - option
 --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:9:17
  |
9 | const _: bool = cmp::const_lt_for!();
  |                 ^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_lt_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed ``; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:10:17
   |
10 | const _: bool = cmp::const_le_for!();
   |                 ^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_le_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed ``; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:11:17
   |
11 | const _: bool = cmp::const_gt_for!();
   |                 ^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_gt_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed ``; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:12:17
   |
12 | const _: bool = cmp::const_ge_for!();
   |                 ^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_ge_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `;`; valid type arguments:
       - slice
       - option
       - range
       - range_inclusive
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:14:17
   |
14 | const _: bool = cmp::const_eq_for!(;);
   |                 ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `cmp::const_eq_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `;`; valid type arguments:
       - slice
       - option
       - range
       - range_inclusive
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:15:17
   |
15 | const _: bool = cmp::const_ne_for!(;);
   |                 ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_eq_for` which comes from the expansion of the macro `cmp::const_ne_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `;`; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:16:17
   |
16 | const _: Orde = cmp::const_cmp_for!(;);
   |                 ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `cmp::const_cmp_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `;`; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:17:17
   |
17 | const _: bool = cmp::const_lt_for!(;);
   |                 ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_lt_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `;`; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:18:17
   |
18 | const _: bool = cmp::const_le_for!(;);
   |                 ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_le_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `;`; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:19:17
   |
19 | const _: bool = cmp::const_gt_for!(;);
   |                 ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_gt_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `;`; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:20:17
   |
20 | const _: bool = cmp::const_ge_for!(;);
   |                 ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_ge_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `abcd`; valid type arguments:
       - slice
       - option
       - range
       - range_inclusive
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:22:17
   |
22 | const _: bool = cmp::const_eq_for!(abcd;);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `cmp::const_eq_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `abcd`; valid type arguments:
       - slice
       - option
       - range
       - range_inclusive
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:23:17
   |
23 | const _: bool = cmp::const_ne_for!(abcd;);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_eq_for` which comes from the expansion of the macro `cmp::const_ne_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `abcd`; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:24:17
   |
24 | const _: Orde = cmp::const_cmp_for!(abcd;);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `cmp::const_cmp_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `abcd`; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:25:17
   |
25 | const _: bool = cmp::const_lt_for!(abcd;);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_lt_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `abcd`; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:26:17
   |
26 | const _: bool = cmp::const_le_for!(abcd;);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_le_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `abcd`; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:27:17
   |
27 | const _: bool = cmp::const_gt_for!(abcd;);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_gt_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected type argument, passed `abcd`; valid type arguments:
       - slice
       - option
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:28:17
   |
28 | const _: bool = cmp::const_ge_for!(abcd;);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the macro `$crate::cmp::const_cmp_for` which comes from the expansion of the macro `cmp::const_ge_for` (in Nightly builds, run with -Z macro-backtrace for more info)

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:30:48
   |
30 | const _: bool = cmp::const_eq_for!(slice; [0u8]);
   |                                                ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_eq_macros.rs
   |
   |         ;$left_slice:expr,
   |                          ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:31:48
   |
31 | const _: bool = cmp::const_ne_for!(slice; [0u8]);
   |                                                ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_eq_macros.rs
   |
   |         ;$left_slice:expr,
   |                          ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:32:49
   |
32 | const _: Orde = cmp::const_cmp_for!(slice; [0u8]);
   |                                                 ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |         ;$left_slice:expr,
   |                          ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:33:48
   |
33 | const _: bool = cmp::const_lt_for!(slice; [0u8]);
   |                                                ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |         ;$left_slice:expr,
   |                          ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:34:48
   |
34 | const _: bool = cmp::const_le_for!(slice; [0u8]);
   |                                                ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |         ;$left_slice:expr,
   |                          ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:35:48
   |
35 | const _: bool = cmp::const_gt_for!(slice; [0u8]);
   |                                                ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |         ;$left_slice:expr,
   |                          ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:36:48
   |
36 | const _: bool = cmp::const_ge_for!(slice; [0u8]);
   |                                                ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/cmp/const_ord_macros.rs
   |
   |         ;$left_slice:expr,
   |                          ^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:38:49
   |
38 | const _: bool = cmp::const_eq_for!(slice; [0u8],);
   |                                                 ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right_slice:expr`
  --> src/cmp/const_eq_macros.rs
   |
   |         $right_slice:expr
   |         ^^^^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:39:49
   |
39 | const _: bool = cmp::const_ne_for!(slice; [0u8],);
   |                                                 ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right_slice:expr`
  --> src/cmp/const_eq_macros.rs
   |
   |         $right_slice:expr
   |         ^^^^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:40:50
   |
40 | const _: Orde = cmp::const_cmp_for!(slice; [0u8],);
   |                                                  ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right_slice:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |         $right_slice:expr
   |         ^^^^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:41:49
   |
41 | const _: bool = cmp::const_lt_for!(slice; [0u8],);
   |                                                 ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right_slice:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |         $right_slice:expr
   |         ^^^^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:42:49
   |
42 | const _: bool = cmp::const_le_for!(slice; [0u8],);
   |                                                 ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right_slice:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |         $right_slice:expr
   |         ^^^^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:43:49
   |
43 | const _: bool = cmp::const_gt_for!(slice; [0u8],);
   |                                                 ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right_slice:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |         $right_slice:expr
   |         ^^^^^^^^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:44:49
   |
44 | const _: bool = cmp::const_ge_for!(slice; [0u8],);
   |                                                 ^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$right_slice:expr`
  --> src/cmp/const_ord_macros.rs
   |
   |         $right_slice:expr
   |         ^^^^^^^^^^^^^^^^^

error[E0601]: `main` function not found in crate `$CRATE`
  --> tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs:44:51
   |
44 | const _: bool = cmp::const_ge_for!(slice; [0u8],);
   |                                                   ^ consider adding a `main` function to `$DIR/tests/misc_tests/cmp_macros_ui_tests/cmp_comparison_for_macros_nocomparator_syntax_err.rs`
